source/CIS/cis-4.rst

```diff
@@ -0,0 +1,796 @@
+The checksum is serialized by 1 byte to indicate whether a hash of the metadata is included.
+If its value is 0, then there is no hash, if the value is 1 then 32 bytes for a SHA256 hash (``hash``) follows::
+
+  MetadataChecksum ::= (0: Byte)
+                     | (1: Byte) (hash: Byte³²)
+
+  MetadataUrl ::= (n: Byte²) (url: Byteⁿ) (checksum: MetadataChecksum)
+
+.. _CIS-4-ContractAddress:
+
+``ContractAddress``
+^^^^^^^^^^^^^^^^^^^
+
+An address of a contract instance.
+It consists of an index and a subindex both unsigned 64-bit integers.
+
+It is serialized as: First 8 bytes for the index (``index``) followed by 8 bytes for the subindex (``subindex``) both little endian::
+
+  ContractAddress ::= (index: Byte⁸) (subindex: Byte⁸)
+
+
+.. _CIS-4-AccountAddress:
+
+``AccountAddress``
+^^^^^^^^^^^^^^^^^^
+
+An address of an account.
+
+It is serialized as 32 bytes::
+
+  AccountAddress ::= (address: Byte³²)
+
+.. _CIS-4-EntrypointName:
+
+``EntrypointName``
+^^^^^^^^^^^^^^^^^^
+
+A name for a smart contract function entrypoint.
+
+It is serialized as: First 2 bytes encode the length (``n``) of the entrypoint name, followed by this many bytes for the entrypoint name (``entrypoint``)::
+
+  EntrypointName ::= (n: Byte²) (entrypoint: Byteⁿ)
+
+.. _CIS-4-Timestamp:
+
+``Timestamp``
+^^^^^^^^^^^^^
+
+A timestamp given in milliseconds since unix epoch.
+It consists of an unsigned 64-bit integer.
+
+It is serialized as 8 bytes in little endian::
+
+  Timestamp ::= (milliseconds: Byte⁸)
+
+.. _CIS-4-Nonce:
+
+``Nonce``
+^^^^^^^^^
+
+An unsigned 64-bit integer number that increases sequentially to protect against replay attacks.
+
+It is serialized as 8 bytes in little endian::
+
+  Nonce ::= (nonce: Byte⁸)
+
+.. _CIS-4-PublicKeyEd25519:
+
+``PublicKeyEd25519``
+^^^^^^^^^^^^^^^^^^^^
+
+A public key is represented as a 32-byte array.
+
+It is serialized as 32 bytes::
+
+  PublicKeyEd25519 ::= (key: Byte³²)
+
+.. _CIS-4-SignatureEd25519:
+
+``SignatureEd25519``
+^^^^^^^^^^^^^^^^^^^^
+
+Signature for an Ed25519 message.
+
+It is serialized as 64 bytes::
+
+  SignatureEd25519 ::= (signature: Byte⁶⁴)
+
+.. _CIS-4-SigningData:
+
+``SigningData``
+^^^^^^^^^^^^^^^
+
+Signing data contains a metadata for the signature that is used to check whether the signed message is designated for the right contract and entrypoint, and it is not expired.
+
+It is serialized as :ref:`CIS-4-ContractAddress` (``contract_address``), :ref:`CIS-4-EntrypointName` (``entrypoint``), :ref:`CIS-4-Nonce` (``nonce``), and :ref:`CIS-4-Timestamp` (``timestamp``)::
+
+  SigningData ::= (contract_address: ContractAddress) (entrypoint: EntrypointName) (nonce: Nonce) (timestamp: Timestamp)
+
+.. _CIS-4-SchemaRef:
+
+``SchemaRef``
+^^^^^^^^^^^^^
+
+A URL of the credential schema.
+
+Serialized in the same way as :ref:`CIS-2 MetadataUrl<CIS-2-MetadataUrl>`.
```

Comment:
Do we maybe want to link to the CIS4 metadataURL serialization instead? It seems odd to me that we duplicate the serialized format and then link to the CIS2 version here. Linking to the CIS4 reference also helps the user stay in the document instead of jumping back and forth.

```suggestion
Serialized in the same way as :ref:`CIS-4-MetadataUrl`.
```

Reply:
Yes, you're right, it should be a local reference to `MetadataUrl`. I see, however, that there is a repetition of common data types in several standards. Maybe (later) we can organize the standards as a library with common definitions being some sort of `std`